### PR TITLE
Use ESFormats for tickets, TMDs and views

### PR DESCRIPTION
**Pull request by leoetlino**
This moves everything related to parsing the ticket and TMD formats to ESFormats. All TMD and ticket getters now return a TMDReader or TicketReader (respectively), instead of a buffer that everyone "parsed" using duplicated offsets.

The first commit moves moves the whole ticket code (which was duplicated in DiscIO and ES) to ESFormats.

While the ticket view structures aren't currently used, they will likely be in the future when we implement some missing ioctlvs (and for a future PR which will change ES_Launch to work based on IOS's behaviour).

The second commit moves the TMD parsing and view generation code to ESFormats as well, since we already have a TMDReader there.

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Did you consider that one suggestion from IRC, to define structs (or use the ones we already have) with `offsetof` to get rid of those magic constants?
It's pretty much a +/-0 since its only used in one place, but for readability/documentation purposes it may help others understand the code.

**Comment by delroth** [comment]
I disagree on this being in Common/. Common/ should be for code that is completely independent from GC/Wii emulation. A good guideline would be "would this be useful for PPSSPP? if not, it has no place in Common/".

DiscIO/ being split from Core/ is historical and probably a bad idea at this point.

File: Source/Core/Core/IOS/ES/Formats.cpp [hunk 1]
```diff
@@ -5,17 +5,32 @@
...
 #include "Common/CommonFuncs.h"
 #include "Common/CommonTypes.h"
 
-namespace IOS
+namespace ES
 {
-namespace HLE
+constexpr size_t CONTENT_VIEW_SIZE = 0x10;
+
+std::vector<u8> AESDecode(const u8* key, u8* iv, const u8* src, u32 size)
```
Comment (delroth) [inline Source/Core/Core/IOS/ES/Formats.cpp]: This looks like a function that should be somewhere in Common/

File: Source/Core/DiscIO/NANDContentLoader.cpp [hunk 2]
```diff
@@ -248,92 +200,64 @@ bool CNANDContentLoader::Initialize(const std::string& name)
...
-    std::copy(hash_begin, hash_end, content.m_SHA1Hash);
+    const auto& content = contents.at(i);
 
     if (m_IsWAD)
     {
-      u32 rounded_size = Common::AlignUp(content.m_Size, 0x40);
-
-      iv.fill(0);
-      std::copy(&tmd[entry_offset + 0x01E8], &tmd[entry_offset + 0x01E8 + 2], iv.begin());
+      // The content index is used as IV (2 bytes); the remaining 14 bytes are zeroes.
```
Comment (delroth) [inline Source/Core/DiscIO/NANDContentLoader.cpp]: We do that in a few places -- might be worth having an abstraction for it?

File: Source/Core/DolphinWX/FrameTools.cpp [hunk 3]
```diff
@@ -1207,23 +1207,7 @@ void CFrame::OnInstallWAD(wxCommandEvent& event)
...
-    const char region = sys_menu_loader.GetCountryChar();
-    menu_item->Enable();
-    menu_item->SetItemLabel(wxString::Format(_("Load Wii System Menu %d%c"), version, region));
-  }
-  else
-  {
-    menu_item->Enable(false);
-    menu_item->SetItemLabel(_("Load Wii System Menu"));
-  }
+  GetMenuBar()->Refresh(true, nullptr);
```
Comment (delroth) [inline Source/Core/DolphinWX/FrameTools.cpp]: Is this change related?

**Comment by delroth** [comment]
The only important thing I'm not too happy with yet is ES vs. ::ES. Not completely sure how to resolve this, but it's error-prone and annoying.

**Comment by lioncash** [comment]
Given ES is part of IOS, it should likely remain within the `IOS::HLE` namespace hierarchy like everything else, or at the very least `IOS`.

**Comment by leoetlino** [comment]
Well, I think a small difference with the rest of IOS is that there is no emulation at all; it's all format parsing code which is used outside of IOS as well. But I agree that it's technically still part of IOS, so what do you think about `IOS::ES`?

**Comment by lioncash** [comment]
Also to touch on what @delroth said: using the organization as it currently is will introduce or further cement (if one is already present) a cyclical dependency between Core and DiscIO

**Comment by delroth** [comment]
+1 for IOS::ES.

File: Source/Core/DiscIO/NANDContentLoader.cpp (hunk 2 above)
Comment (leoetlino) [inline Source/Core/DiscIO/NANDContentLoader.cpp]: You mean something to decrypt a content?

File: Source/Core/DolphinWX/FrameTools.cpp (hunk 3 above)
Comment (leoetlino) [inline Source/Core/DolphinWX/FrameTools.cpp]: Somewhat, yes -- I'd have to implement my changes twice because this function is duplicated.

File: Source/Core/Core/IOS/ES/Formats.cpp (hunk 1 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/Formats.cpp]: Moved.

**Comment by leoetlino** [comment]
Fixed the namespace and moved the AES function to Common/Crypto. Not quite sure how to fix the dependency issue though -- I think ESFormats definitely belongs to IOS/ES, but the only way to fix it is to move it to DiscIO. @delroth, are you okay with that?

**Comment by leoetlino** [comment]
Looks like this broke region detection for the system menu. Looking into it.

**Comment by leoetlino** [comment]
For the reference, the plan was to eventually move the other ES specific parts to ES, and renaming DiscIO (and potentially moving it?) so that its name more accurately describes what it does; it handles more than just disc files (e.g. NAND files).

The cyclical dependency will be removed in a followup PR, not here because it would require the DiscIO code to be reorganised right now, which would be out of this PR's scope (imo) and heavily conflict with this PR.

re part of the content decrypt function being duplicated, I have a PR ready to reuse more code for the content adding code, so the content IV stuff would not be duplicated anymore -- I don't think it should be part of this PR, though.

Is there anything else that should be changed, or can this be merged?

**Comment by JMC47** [comment]
Is there anything else waiting on this, I'd like to get this and the one that relies on it merged so the disc channel can work again.
